flang/lib/Lower/OpenMP.cpp

```diff
@@ -87,19 +88,35 @@ genOMP(Fortran::lower::AbstractConverter &absConv,
     auto &firOpBuilder = absConv.getFirOpBuilder();
     auto currentLocation = absConv.getCurrentLocation();
     auto insertPt = firOpBuilder.saveInsertionPoint();
+
+    // Clauses.
+    mlir::Value numThreads;
+
+    const auto &parallelOpClauseList =
+        std::get<Fortran::parser::OmpClauseList>(blockDirective.t);
+    for (const auto &clause : parallelOpClauseList.v) {
+      if (const auto &numThreadsClause =
+              std::get_if<Fortran::parser::OmpClause::NumThreads>(&clause.u)) {
+        numThreads = firOpBuilder.createIntegerConstant(
+            currentLocation, firOpBuilder.getIntegerType(32),
+            Fortran::evaluate::ToInt64(
```

Comment:
Will this evaluate work fine if num_threads is a variable? Can you add a test?

Reply:
My guess is that you have to lower this to a Value since `num_thread` is not a `scalarIntConstantExpr` but just a `scalarIntExpr`. I guess you took example on the `collapse` clause in my last PR but collapse is a `scalarIntConstantExpr` so we can evaluate it to an integer but not `num_thread`

Reply:
Hi Thanks for review!
Feel free to merge and try in your local repos.
1. There are issues WRT num_threads I create this PR to bring all this on table and discuss.
> Will this evaluate work fine if num_threads is a variable? Can you add a test?
Consider below snippet:
```
integer alpha
alpha = 2
!$OMP PARALLEL NUM_THREADS(alpha)
```
This snippet causes crashing of `bbc` itself, at the `pft` creation level. (Should we file an issue ?)
```
bbc parallel.f90 -fopenmp -fdebug-dump-pre-fir
```
> My guess is that you have to lower this to a Value since num_thread is not a scalarIntConstantExpr but just a scalarIntExpr. ....
2.  Okay `pft` dump clears that(for the test case in this PR):
```
  OmpClauseList -> OmpClause -> NumThreads -> Scalar -> Integer -> Expr
```
I think saw `Constant` anyway thanks I'll correct this

3.  Last issue, after this code-change parallel directive without `Num_thread` clause causes assertion:
```
!$OMP PARALLEL
```
Reason behind this is that, `mlir::ValueRange` dereference a nulltype Line  no.109
```
mlir::ValueRange range(numThreads); 
```
When the clause is not present, a nulltype gets created and inserted in this range. This maybe avoided with if-else but for longer term we need a better interface, since in the long term problem here is that every argument is `optional`.